### PR TITLE
all: initialize atlas versioned migration

### PR DESCRIPTION
**Pull request by datdao**
### Context
This PR is related to supporting the `Versioned Migration` feature for the Operator. It is the first pull request (PR) that generates an API template without any implementation inside the controller

###  Changes
Most of the files are new and were generated from operator-sdk using the command `operator-sdk create api --group db --version v1alpha1 --kind AtlasMigration --resource --controller`. Additionally, some files were modified by the command `make generate manifests chart-manifest`.

To conduct a quick review, please focus on these two files:  
- `api/v1alpha1/atlasmigration_types.go`, This file defines the specification for the AtlasMigration resource
- `controllers/atlasmigration_controller.go`, This file defines the logic to implement when the resource is changed

By default, The controller will be marked as `experiment` feature, to enable it we need to config env at deployment.yaml of atlas operator

```
    spec:
      containers:
        - name: manager
          image: >-
            controller:f72be041b0a0305ad6deb703e0d8a191e1fd4cae312b872cfcd1229d8c288ab9
          env:
            - name: 'EXPERIMENTAL'
              value: 'true'
          command:
            - /manager
          ...
```

### CODE REVIEW

File: api/v1alpha1/atlasmigration_types.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	// URLs may be defined as a secret key reference.
+	URLFrom URLFrom `json:"urlFrom,omitempty"`
+	// Cloud defines the Atlas Cloud configuration.
+	Cloud Cloud `json:"cloud,omitempty"`
+	// Desired Version of the target database schema.
+	Version string `json:"version,omitempty"`
+	// Exclude a list of glob patterns used to filter existing resources being taken into account.
+	Exclude []string `json:"exclude,omitempty"`
+	// Policy defines the policies to apply when managing the schema change lifecycle.
+	Policy Policy `json:"policy,omitempty"`
```
Comment (rotemtam) [inline api/v1alpha1/atlasmigration_types.go:40]:
```suggestion
```
Let's consider support for these later

File: controllers/atlasmigration_controller.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	}()
+
+	// Skip if atlas version already up to date
+	if am.Status.LastAppliedVersion == am.Spec.Version {
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
```
Comment (rotemtam) [inline controllers/atlasmigration_controller.go:111]: if clientErr?

File: api/v1alpha1/atlasmigration_types.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	// URL of the target database schema.
+	URL string `json:"url,omitempty"`
+	// URLs may be defined as a secret key reference.
+	URLFrom URLFrom `json:"urlFrom,omitempty"`
+	// Cloud defines the Atlas Cloud configuration.
+	Cloud Cloud `json:"cloud,omitempty"`
+	// Desired Version of the target database schema.
+	Version string `json:"version"`
+	// DirFrom defines the directory to use for migrations as a configmap key reference.
+	Dir Dir `json:"dir"`
```
Comment (giautm) [inline api/v1alpha1/atlasmigration_types.go:38]:
```suggestion
	// Dir defines the directory to use for migrations as a configmap key reference.
	Dir Dir `json:"dir"`
```

File: api/v1alpha1/atlasmigration_types.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
```
Comment (giautm) [inline api/v1alpha1/atlasmigration_types.go:25]:
```suggestion
```

File: api/v1alpha1/atlasmigration_types.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+// Dir defines the place where migrations are stored.
+type Dir struct {
+	// ConfigMapRef defines the configmap to use for migrations
+	ConfigMapRef string `json:"configMapRef,omitempty"`
+	// Remote defines the Atlas Cloud directory migration.
+	Remote Remote `json:"remote,omitempty"`
+}
+
+// RemoteDir defines the Atlas Cloud directory migration.
```
Comment (giautm) [inline api/v1alpha1/atlasmigration_types.go:56]:
```suggestion
// Remote defines the Atlas Cloud directory migration.
```

File: api/v1alpha1/atlasmigration_types.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// AtlasMigrationStatus defines the observed state of AtlasMigration
+type AtlasMigrationStatus struct {
+	// Conditions represent the latest available observations of an object's state.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// LastAppliedVersion is the version of the most recent successful versioned migration.
+	LastAppliedVersion string `json:"lastAppliedVersion,omitempty"`
+	//LastRunURL is the Deployment URL of the most recent successful versioned migration.
+	LastRunURL string `json:"lastRunUrl,omitempty"`
```
Comment (giautm) [inline api/v1alpha1/atlasmigration_types.go:80]:
```suggestion
	//LastDeploymentURL is the Deployment URL of the most recent successful versioned migration.
	LastDeploymentURL string `json:"lastDeploymentUrl,omitempty"`
```

Comment (giautm) [reply api/v1alpha1/atlasmigration_types.go:80]: why not use same term?

File: api/v1alpha1/atlasmigration_types.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+type Cloud struct {
+	URL       string    `json:"url,omitempty"`
+	TokenFrom TokenFrom `json:"tokenFrom,omitempty"`
+	Project   string    `json:"project,omitempty"`
+}
+
+// Dir defines the place where migrations are stored.
+type Dir struct {
+	// ConfigMapRef defines the configmap to use for migrations
+	ConfigMapRef string `json:"configMapRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (datdao) [reply api/v1alpha1/atlasmigration_types.go:48]: `corev1.ConfigMapKeySelector` This struct contains info to retrieve a value with the corresponding key, not the whole object

File: controllers/atlasmigration_controller.go (hunk 2 above)
Comment (datdao) [reply controllers/atlasmigration_controller.go]: This logic was moved to the next PR, we will discuss on that

File: config/crd/patches/cainjection_in_atlasmigrations.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
+apiVersion: apiextensions.k8s.io/v1
```
Comment (rotemtam) [inline config/crd/patches/cainjection_in_atlasmigrations.yaml:2]: I think this isn't needed

File: config/crd/patches/webhook_in_atlasmigrations.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# The following patch enables a conversion webhook for the CRD
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
```
Comment (rotemtam) [inline config/crd/patches/webhook_in_atlasmigrations.yaml:3]: same

File: controllers/atlasmigration_controller.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the AtlasMigration object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
```
Comment (rotemtam) [inline controllers/atlasmigration_controller.go:45]:
```suggestion
```

File: controllers/atlasmigration_controller.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
```
Comment (rotemtam) [inline controllers/atlasmigration_controller.go:48]:
```suggestion
```

File: controllers/suite_test.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2023.
```
Comment (rotemtam) [inline controllers/suite_test.go:2]: Remove this file as well

File: main.go [hunk 13]
```diff
@@ -112,6 +112,13 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "AtlasSchema")
 		os.Exit(1)
 	}
+	if err = (&controllers.AtlasMigrationReconciler{
```
Comment (rotemtam) [inline main.go:115]: Let's gate this with an env var / flag (off by default) until the feature is ready.

Comment (rotemtam) [reply main.go:115]: i think it will take a few PRs to get the feature out and I prefer we can test it while not exposing it to users yet

Comment (datdao) [reply main.go:115]: ok I will gate it with env,
```
	experimental := os.Getenv("EXPERIMENTAL") != ""
	if experimental {
		setupLog.Info("Experiment mode is enabled")
		if err = (&controllers.AtlasMigrationReconciler{
			Client: mgr.GetClient(),
			Scheme: mgr.GetScheme(),
		}).SetupWithManager(mgr); err != nil {
			setupLog.Error(err, "unable to create controller", "controller", "AtlasMigration")
			os.Exit(1)
		}
	}
```

**Review by rotemtam — APPROVED** [review]
LGTM , please remove suite test file first

**Pull request merged**